### PR TITLE
Contribute ABAP Code Review Guidelines

### PR DESCRIPTION
**Pull request by xtough**
proposal of a separate guide from @larshp and myself.

### CODE REVIEW

**Comment by xtough** [comment]
@larshp FYI

**Comment by larshp** [comment]
PDF build: https://github.com/xtough/styleguides/releases/download/latest/abap-code-review-guide.pdf

File: abap-code-review/README.md
Comment (KlausHaeuptleSAP) [inline abap-code-review/README.md:4]: Why is an additional license needed?

File: abap-code-review/README.md
Comment (KlausHaeuptleSAP) [inline abap-code-review/README.md:14]: Could you add a few sentences why AsciiDoc has been chosen instead of Markdown.

File: abap-code-review/README.md
Comment (KlausHaeuptleSAP) [inline abap-code-review/README.md:6]: Is the PDF the main reading or just an alternative format?

File: abap-code-review/README.md
Comment (larshp) [reply abap-code-review/README.md:4]: the link is to the existing license, https://github.com/xtough/styleguides/blob/main/abap-code-review/README.md

File: abap-code-review/README.md
Comment (larshp) [reply abap-code-review/README.md:6]: well, both can be read, just the diagrams are only rendered in the PDF

File: abap-code-review/README.md
Comment (larshp) [reply abap-code-review/README.md:14]: updated, https://github.com/xtough/styleguides/commit/383e88c89ed976a8ec4dc20dd24e4176a722cc6c

Comment (larshp) [reply abap-code-review/README.md:14]: also see https://github.com/SAP/styleguides/issues/5

File: abap-code-review/README.md
Comment (xtough) [reply abap-code-review/README.md:4]: ...in fact to https://github.com/xtough/styleguides/blob/main/LICENSE, which would become https://github.com/SAP/styleguides/blob/main/LICENSE after merging the PR.

File: abap-code-review/src/code_review.adoc
Comment (KlausHaeuptleSAP) [inline abap-code-review/src/code_review.adoc:30]: Should we also add code-owners for the code review section?

I think it will also become relevant when we add Clean UI5.

File: abap-code-review/src/code_review.adoc
Comment (KlausHaeuptleSAP) [inline abap-code-review/src/code_review.adoc:42]: Sometimes you also want to have a different perspective. So an additional code review with different expertise or maturity can help.

File: abap-code-review/src/code_review.adoc
Comment (larshp) [reply abap-code-review/src/code_review.adoc:30]: as infrastructure for the "styleguide" repository? Yea, I think so, but suggest moving it to a different PR/issue

File: abap-code-review/README.md
Comment (KlausHaeuptleSAP) [inline abap-code-review/README.md:2]: Should some sentences be added that the setup there are currently some investments done and the setup will probably change in the coming year?

Comment (larshp) [reply abap-code-review/README.md:2]: ??? so SAP is moving in a different direction, which is not the one described??

File: abap-code-review/src/code_review.adoc
Comment (JozsefSzikszai) [reply abap-code-review/src/code_review.adoc:30]: "Manual selection happens either by the author (...), or by interested reviewers"
 I have worked in an organization, where the team leader who assigned the task to a specific developer, also decided who is doing the code review.

File: abap-code-review/README.md
Comment (xtough) [reply abap-code-review/README.md:2]: This guide is independent of any ongoing discussions on technical foundations how to add code reviews, e.g., to classical CTS flows. --> no need to add confusion here

File: abap-code-review/src/code_review.adoc
Comment (xtough) [reply abap-code-review/src/code_review.adoc:30]: > "Manual selection happens either by the author (...), or by interested reviewers"
> I have worked in an organization, where the team leader who assigned the task to a specific developer, also decided who is doing the code review.

Sounds not very agile to me, but, yes, this could be covered as well

Comment (xtough) [reply abap-code-review/src/code_review.adoc:30]: > Should we also add code-owners for the code review section?
> 
> I think it will also become relevant when we add Clean UI5.

We can add this later, once the initial PR is merged

File: abap-code-review/src/code_review.adoc
Comment (xtough) [reply abap-code-review/src/code_review.adoc:42]: added this aspect to motivation subsection above

File: abap-code-review/src/code_review.adoc
Comment (xtough) [reply abap-code-review/src/code_review.adoc:30]: done with https://github.com/xtough/styleguides/commit/72e265847b90c07f51da3060796424ac11c84bf8

**Comment by xtough** [comment]
I would suggest to concentrate on grave problems with the initial commit and tackle individual smaller points later via issues.

**Review by HrFlorianHoffmann — APPROVED** [review]

**Review by KlausHaeuptleSAP — APPROVED** [review]

**Comment by jfilak** [comment]
Nice, I will try to review tomorrow morning.

File: abap-code-review/src/code_review.adoc
Comment (filak-sap) [inline abap-code-review/src/code_review.adoc:20]: Perhaps, you can use some advanced text editor like vim and reformat to 80-120 characters per line.
That helps pointing out problems in the text the same way as in source codes.

There is a missing space:
```
earlier time.The main
```

File: abap-code-review/src/code_review.adoc
Comment (filak-sap) [inline abap-code-review/src/code_review.adoc:22]: This is false claim because no ABAPists ever released a "small" transport. So, I would say:

```
An *ABAP Transport Request* fulfills these criteria in theory but the common practice of optimizing transports amount for less transports over time causes unnecessary growing of transport size.
```

File: abap-code-review/src/code_review.adoc
Comment (filak-sap) [inline abap-code-review/src/code_review.adoc:25]: Also, if you open a PR it's always a good idea to have *sane* amount of self contained commits. This PR is an example of how it should not look like.

File: abap-code-review/src/code_review.adoc
Comment (filak-sap) [inline abap-code-review/src/code_review.adoc:36]: I sometimes review foreign PRs just out of curiosity and for learning purposes.

File: abap-code-review/src/introduction.adoc
Comment (filak-sap) [inline abap-code-review/src/introduction.adoc:27]: Are you sure you want to call these code analysis steps "Automated Review"? The word Review somehow sounds inappropriate here.

File: abap-code-review/src/introduction.adoc
Comment (filak-sap) [inline abap-code-review/src/introduction.adoc:37]: I don't think a source repository supports validation but source code management system do:
https://de.wikipedia.org/wiki/Source_Code_Control_System

File: abap-code-review/src/introduction.adoc
Comment (filak-sap) [inline abap-code-review/src/introduction.adoc:38]: I would replace "Git platforms" with "Git based collaboration platforms" .

File: abap-code-review/src/scenarios.adoc
Comment (filak-sap) [inline abap-code-review/src/scenarios.adoc:38]: The pull to target systems can be automated.

File: abap-code-review/src/scenarios.adoc
Comment (filak-sap) [inline abap-code-review/src/scenarios.adoc:40]: We do push using abapgit from 3 different AS ABAP systems into shared repositories. Why don't you allow such a possibility? Especially abapGit is a perfect example of such a setup.

File: abap-code-review/src/tools.adoc
Comment (filak-sap) [inline abap-code-review/src/tools.adoc:21]: Yes, commits are the most important features and thus one should learn how to make good commits:
- https://chris.beams.io/posts/git-commit/

The most valuable asset of git repositories is the "history" and without good commit messages, the history is useless.

**Comment by filak-sap** [comment]
Would it be possible to place a paragraph about the need to use "good commit" messages and the need to "amend or rebase review branches" before sending them for review?

File: abap-code-review/src/code_review.adoc
Comment (larshp) [reply abap-code-review/src/code_review.adoc]: thanks, added in https://github.com/SAP/styleguides/pull/183/commits/2cbdd63fbe659fc8d41f6799b847a1512b60ed69

File: abap-code-review/src/scenarios.adoc
Comment (larshp) [reply abap-code-review/src/scenarios.adoc:38]: with code reviews in mind, the target system would be a system where development happens, I guess its typically not desired to pull automatically into these systems, as development can be ongoing?

File: abap-code-review/src/tools.adoc
Comment (larshp) [reply abap-code-review/src/tools.adoc:21]: yea, suggest a contribution regarding commit messages for the "Best Practices" chapter after the PR has been merged

File: abap-code-review/src/code_review.adoc
Comment (xtough) [reply abap-code-review/src/code_review.adoc:22]: The truth is probably somewhere in the middle, depending on many factors. I suggest to tread this as a separate issue once the initial PR is merged.

File: abap-code-review/src/code_review.adoc
Comment (xtough) [reply abap-code-review/src/code_review.adoc:25]: Number of commits is debatable. If you squash upon merging PRs, individual commits become irrelevant. The important point is the number of changed lines, IMHO.

File: abap-code-review/src/code_review.adoc
Comment (xtough) [reply abap-code-review/src/code_review.adoc:36]: IMHO, that is already covered in the introductory remarks to this section.

File: abap-code-review/src/introduction.adoc
Comment (xtough) [reply abap-code-review/src/introduction.adoc]: Done.  Changed to "Automated Checks"

File: abap-code-review/src/introduction.adoc
Comment (xtough) [reply abap-code-review/src/introduction.adoc]: Done. VCS is another term for the same.

File: abap-code-review/src/introduction.adoc
Comment (xtough) [reply abap-code-review/src/introduction.adoc]: done.

File: abap-code-review/src/scenarios.adoc
Comment (xtough) [reply abap-code-review/src/scenarios.adoc:38]: Sounds debatable to me. Suggest to tread this in a separate issue/PR once the initial PR is merged.

File: abap-code-review/src/scenarios.adoc
Comment (xtough) [reply abap-code-review/src/scenarios.adoc:40]: Pushing the same package tree to different repos sounds strange to me. 
I suggest discuss this in a separate issue after the initial PR is merged.

File: abap-code-review/src/tools.adoc
Comment (xtough) [reply abap-code-review/src/tools.adoc:21]: ...besides, individual commits become irrelevant if you squash them upon merging PRs. See my previous commit above.

File: abap-code-review/src/code_review.adoc
Comment (xtough) [reply abap-code-review/src/code_review.adoc:25]: Regarding the PR at hand, you are of course right. On the other hand, we had to make some sacrifices to efficiency. Collaboratively writing 2000+ lines among the 2 of us was already hard enough ;-)

**Comment by xtough** [comment]
> Would it be possible to place a paragraph about the need to use "good commit" messages and the need to "amend or rebase review branches" before sending them for review?

@filak-sap Valid points. However, I would suggest handling both in separate issues after the initial PR is merged. 
* If PRs are squashed after merging, formulating and formatting them becomes more important than individual commit messages. Same rules as for commits apply (e.g., https://chris.beams.io/posts/git-commit/ )
* While amending commits is required in Gerrit, it can even be dangerous in GitHub and other platforms. Rebasing vs. Merging parent branches into feature branches is also a matter of philosophy. ...You see that this point also requires some discussions.

**Comment by jfilak** [comment]
> If PRs are squashed after merging,

Correct from the history point of view, but even reviewers have simpler life if you split your changes into small understandable units.

> While amending commits is required in Gerrit, it can even be dangerous in GitHub

Amending in Gerrit was taken to the absurd state and has nothing common with amending I am asking for.
We can talk about amending and rebasing a lot but in GitHub with forks and private branches, these are not problem at all.
One can merge a PR with the merge strategy but at least, the PR branch should be rebased to the main branch first.

However, I didn't want to talk about merging PRs but about creating PRs with good commits, so the human reviewer can see the flow of changes and always can focus on a minimal set of changes.

**Comment by xtough** [comment]
> However, I didn't want to talk about merging PRs but about creating PRs with good commits, so the human reviewer can see the flow of changes and always can focus on a minimal set of changes.

This is what I meant with "debatable".  😋
I personally don't care how many local commits it took a contributor to come up with a PR, as long as the latter is in itself concise and "reviewable".

As I wrote, I suggest to discuss this matter in a separate issue once the initial PR is merged.

File: abap-code-review/src/scenarios.adoc
Comment (jfilak) [reply abap-code-review/src/scenarios.adoc:40]: I was under the impression that you too have a local git repo made of this repo. We do the same, we have several private AS ABAP systems and all working on the same repo.

**Review by filak-sap — APPROVED** [review]
Well done! I see no blockers for this PR to land in the main branch.

**Pull request merged**
